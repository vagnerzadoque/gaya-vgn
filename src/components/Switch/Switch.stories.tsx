import React, { useState } from 'react'
import type { Meta, StoryObj } from '@storybook/react'
import { fn } from '@storybook/test'
import { Switch, SwitchProps } from '.'
import StoryContainer from '../../helpers/StoryContainer'

const componentStatus = `
> Switches allow users to enable or disable an option.

## Properties
| Property                      | Values                                        |    Status           |
|---                            |                                            ---|                  ---|
| **variant (no prop)**         | Standard                                      | ✅ Available        |
| **checked**                   | true/false                                    | ✅ Available        |
| **value**                     | string                                        | ✅ Available        |
| **id**                        | string                                        | ✅ Available        |
| **onChange**                  | function                                      | ✅ Available        |
| **disabled**                  | true/false                                    | ✅ Available        |
| **brand**                     | avon, avon_v2, natura, natura_v2, theBodyShop, <br /> consultoriaDeBeleza, casaEestilo            | ✅ Available        |

## Technical Usages Examples
`

const meta: Meta<SwitchProps> = {
  title: 'Components/Switch',
  component: Switch,
  parameters: {
    componentSubtitle: '',
    docs: { description: { component: componentStatus } }
  }
};

export default meta;
type Story = StoryObj<SwitchProps>;

export const Playground: Story = {
  render: ({ checked, ...args }) => {
    const [check, setCheck] = useState(checked)

    return <Switch {...args} checked={check} onChange={() => setCheck(!check)} />
  }
}

export const Disabled: Story = {
  args: {
    disabled: true
  },
  render: (args) => (
    <StoryContainer>
      <Switch {...args} checked={false} id="firstSwitch" />
      <Switch {...args} checked id="secondSwitch" />
    </StoryContainer>
  )
}
