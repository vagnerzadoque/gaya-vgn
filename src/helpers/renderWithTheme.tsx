import React from 'react'
import { JssProvider, SheetsRegistry } from 'react-jss'
import themes from '@naturacosmeticos/natds-themes'
import { render } from '@testing-library/react'
import { ThemeProvider } from '../ThemeProvider'

const renderWithTheme = (component: JSX.Element) => {
  const sheets = new SheetsRegistry()

  const renderTree = render(
    <JssProvider registry={sheets}>
      <ThemeProvider theme={themes.natura.light}>
        {component}
      </ThemeProvider>
    </JssProvider>
  )

  return { styles: sheets.toString(), component: renderTree }
}

export default renderWithTheme
