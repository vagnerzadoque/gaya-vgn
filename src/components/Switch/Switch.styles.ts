/* eslint-disable no-nested-ternary */
/* eslint-disable max-len */
import { createUseStyles } from 'react-jss'
import { Theme } from '@naturacosmeticos/natds-themes'
import { SwitchProps } from './Switch.props'
import { buildTheme } from '../../ThemeProvider'

type SwitchStyleProps = Pick<SwitchProps, 'checked' | 'disabled' | 'brand'>

const transitionProps = '150ms cubic-bezier(0.4, 0, 0.2, 1) 0ms'

const getThemeSw = (theme: Theme) => ({ brand }: SwitchStyleProps) => {
  const themeSelectFocus = buildTheme(brand, 'light')

  if (brand) {
    return themeSelectFocus.color.primary
  }
  return theme.color.primary
}

const styles = createUseStyles((theme: Theme) => ({
  switchContainer: {
    width: 56,
    height: 40,
    position: 'relative',
    display: 'flex'
  },
  switchBase: {
    display: 'flex',
    alignItems: 'center',
    cursor: ({ disabled }: SwitchStyleProps) => (disabled ? 'default' : 'pointer'),
    pointerEvents: ({ disabled }: SwitchStyleProps) => (disabled ? 'none' : 'unset'),
    width: '100%',
    height: '100%',
    '&:hover $switchOverlay': {
      opacity: theme.opacity.veryLow,
      backgroundColor: ({ checked, brand }: SwitchStyleProps) => (checked ? getThemeSw(theme)({ brand }) : theme.color.highlight)
    }
  },
  switchInput: {
    position: 'absolute',
    opacity: 0,
    '&:focus + $switchBase $switchOverlay': {
      opacity: theme.opacity.low,
      backgroundColor: ({ checked, brand }: SwitchStyleProps) => (checked ? getThemeSw(theme)({ brand }) : theme.color.highlight)
    }
  },
  switchTrack: {
    transition: `background-color ${transitionProps}`,
    left: 10,
    position: 'absolute',
    width: 36,
    height: 14,
    borderRadius: 20,
    backgroundColor: ({ checked, disabled, brand }: SwitchStyleProps) => (disabled
      ? theme.color.lowEmphasis
      : checked
        ? getThemeSw(theme)({ brand })
        : theme.color.mediumEmphasis),
    opacity: ({ checked, disabled }: SwitchStyleProps) => (disabled || !checked ? 1 : theme.opacity.medium)
  },
  switchOverlay: {
    left: ({ checked }: SwitchStyleProps) => (checked ? 16 : 0),
    transition: `left ${transitionProps}`,
    position: 'absolute',
    width: 40,
    height: 40,
    borderRadius: '50%'
  },
  switchThumb: {
    left: ({ checked }: SwitchStyleProps) => (checked ? 26 : 10),
    transition: `left ${transitionProps}`,
    position: 'absolute',
    width: 20,
    height: 20,
    boxShadow: '0px 1px 3px #00000036',
    borderRadius: '50%',
    backgroundColor: ({ checked, disabled, brand }: SwitchStyleProps) => (disabled || !checked ? theme.color.surface : getThemeSw(theme)({ brand }))
  }
}))

export default styles
