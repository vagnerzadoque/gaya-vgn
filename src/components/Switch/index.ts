import Switch from './Switch'

export { Switch }
export { SwitchProps } from './Switch.props'
export default Switch
