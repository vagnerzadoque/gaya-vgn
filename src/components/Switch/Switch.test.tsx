import React from 'react'
import userEvent from '@testing-library/user-event'
import { fireEvent } from '@testing-library/react'
import Switch, { SwitchProps } from '.'
import renderWithTheme from '../../helpers/renderWithTheme'

const switchProps: SwitchProps = {
  onChange: jest.fn(),
  value: 'test'
}

describe('Switch component', () => {
  it('should render correctly when disabled is false and checked is false', () => {
    const { styles, component } = renderWithTheme(<Switch {...switchProps} />)

    expect([styles.toString(), component.container]).toMatchSnapshot()
  })

  it('should render correctly when disabled is false and checked is true', () => {
    const { styles, component } = renderWithTheme(<Switch {...switchProps} checked />)

    expect([styles.toString(), component.container]).toMatchSnapshot()
  })

  it('should render correctly when disabled is true and checked is false', () => {
    const { styles, component } = renderWithTheme(<Switch {...switchProps} disabled />)

    expect([styles.toString(), component.container]).toMatchSnapshot()
  })

  it('should render correctly when disabled is true and checked is true', () => {
    const { styles, component } = renderWithTheme(<Switch {...switchProps} disabled checked />)

    expect([styles.toString(), component.container]).toMatchSnapshot()
  })

  it('should call onChange when click on the switch', () => {
    const onChangeMock = jest.fn()
    const {
      component: { getByTestId }
    } = renderWithTheme(<Switch {...switchProps} onChange={onChangeMock} testID="switch-test" />)

    const switchComponent = getByTestId('switch-test')
    // Using fireEvent instead of userEvent for more direct event triggering
    fireEvent.click(switchComponent)

    expect(onChangeMock).toHaveBeenCalled()
  })

  it('should not call onChange when click on the disabled switch', () => {
    const onChangeMock = jest.fn()
    const {
      component: { getByTestId }
    } = renderWithTheme(
      <Switch {...switchProps} onChange={onChangeMock} testID="switch-test" disabled />
    )

    const switchComponent = getByTestId('switch-test')
    userEvent.click(switchComponent)

    expect(switchComponent).toBeDisabled()
    expect(onChangeMock).not.toHaveBeenCalled()
  })

  it('should blur the switch component when mouse leave', () => {
    const {
      component: { getByTestId, container }
    } = renderWithTheme(<Switch {...switchProps} testID="switch-test" />)

    const switchComponent = getByTestId('switch-test')
    switchComponent.focus()

    expect(switchComponent).toHaveFocus()

    const labelElement = container.querySelector('label')
    fireEvent.mouseLeave(labelElement as Element)

    expect(switchComponent).not.toHaveFocus()
  })
})
