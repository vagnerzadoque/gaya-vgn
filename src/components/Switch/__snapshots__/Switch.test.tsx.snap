// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch component should render correctly when disabled is false and checked is false 1`] = `
[
  ".switchContainer-0-2-1 {
  width: 56px;
  height: 40px;
  display: flex;
  position: relative;
}
.switchBase-0-2-2 {
  width: 100%;
  height: 100%;
  display: flex;
  align-items: center;
}
.switchBase-0-2-2:hover .switchOverlay-0-2-5 {
  opacity: 0.08;
}
.switchInput-0-2-3 {
  opacity: 0;
  position: absolute;
}
.switchInput-0-2-3:focus + .switchBase-0-2-2 .switchOverlay-0-2-5 {
  opacity: 0.12;
}
.switchTrack-0-2-4 {
  left: 10px;
  width: 36px;
  height: 14px;
  position: absolute;
  transition: background-color 150ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  border-radius: 20px;
}
.switchOverlay-0-2-5 {
  width: 40px;
  height: 40px;
  position: absolute;
  transition: left 150ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  border-radius: 50%;
}
.switchThumb-0-2-6 {
  width: 20px;
  height: 20px;
  position: absolute;
  box-shadow: 0px 1px 3px #00000036;
  transition: left 150ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  border-radius: 50%;
}
.switchBase-d0-0-2-7 {
  cursor: pointer;
  pointer-events: unset;
}
.switchBase-d0-0-2-7:hover .switchOverlay-0-2-5 {
  background-color: #111111;
}
.switchInput-d1-0-2-8 {}
.switchInput-d1-0-2-8:focus + .switchBase-0-2-2 .switchOverlay-0-2-5 {
  background-color: #111111;
}
.switchTrack-d2-0-2-9 {
  background-color: #737373;
  opacity: 1;
}
.switchOverlay-d3-0-2-10 {
  left: 0;
}
.switchThumb-d4-0-2-11 {
  left: 10px;
  background-color: #ffffff;
}",
  <div>
    <span
      class=" switchContainer-0-2-1 Switch-gaya"
    >
      <input
        class="switchInput-0-2-3 switchInput-d1-0-2-8"
        id="toggleSwitch"
        type="checkbox"
        value="test"
      />
      <label
        class="switchBase-0-2-2 switchBase-d0-0-2-7"
        for="toggleSwitch"
      >
        <span
          class="switchTrack-0-2-4 switchTrack-d2-0-2-9"
        />
        <span
          class="switchOverlay-0-2-5 switchOverlay-d3-0-2-10"
        />
        <span
          class="switchThumb-0-2-6 switchThumb-d4-0-2-11"
        />
      </label>
    </span>
  </div>,
]
`;

exports[`Switch component should render correctly when disabled is false and checked is true 1`] = `
[
  ".switchContainer-0-2-1 {
  width: 56px;
  height: 40px;
  display: flex;
  position: relative;
}
.switchBase-0-2-2 {
  width: 100%;
  height: 100%;
  display: flex;
  align-items: center;
}
.switchBase-0-2-2:hover .switchOverlay-0-2-5 {
  opacity: 0.08;
}
.switchInput-0-2-3 {
  opacity: 0;
  position: absolute;
}
.switchInput-0-2-3:focus + .switchBase-0-2-2 .switchOverlay-0-2-5 {
  opacity: 0.12;
}
.switchTrack-0-2-4 {
  left: 10px;
  width: 36px;
  height: 14px;
  position: absolute;
  transition: background-color 150ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  border-radius: 20px;
}
.switchOverlay-0-2-5 {
  width: 40px;
  height: 40px;
  position: absolute;
  transition: left 150ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  border-radius: 50%;
}
.switchThumb-0-2-6 {
  width: 20px;
  height: 20px;
  position: absolute;
  box-shadow: 0px 1px 3px #00000036;
  transition: left 150ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  border-radius: 50%;
}
.switchBase-d0-0-2-7 {
  cursor: pointer;
  pointer-events: unset;
}
.switchBase-d0-0-2-7:hover .switchOverlay-0-2-5 {
  background-color: #f4ab34;
}
.switchInput-d1-0-2-8 {}
.switchInput-d1-0-2-8:focus + .switchBase-0-2-2 .switchOverlay-0-2-5 {
  background-color: #f4ab34;
}
.switchTrack-d2-0-2-9 {
  background-color: #f4ab34;
  opacity: 0.48;
}
.switchOverlay-d3-0-2-10 {
  left: 16px;
}
.switchThumb-d4-0-2-11 {
  left: 26px;
  background-color: #f4ab34;
}",
  <div>
    <span
      class=" switchContainer-0-2-1 Switch-gaya"
    >
      <input
        checked=""
        class="switchInput-0-2-3 switchInput-d1-0-2-8"
        id="toggleSwitch"
        type="checkbox"
        value="test"
      />
      <label
        class="switchBase-0-2-2 switchBase-d0-0-2-7"
        for="toggleSwitch"
      >
        <span
          class="switchTrack-0-2-4 switchTrack-d2-0-2-9"
        />
        <span
          class="switchOverlay-0-2-5 switchOverlay-d3-0-2-10"
        />
        <span
          class="switchThumb-0-2-6 switchThumb-d4-0-2-11"
        />
      </label>
    </span>
  </div>,
]
`;

exports[`Switch component should render correctly when disabled is true and checked is false 1`] = `
[
  ".switchContainer-0-2-1 {
  width: 56px;
  height: 40px;
  display: flex;
  position: relative;
}
.switchBase-0-2-2 {
  width: 100%;
  height: 100%;
  display: flex;
  align-items: center;
}
.switchBase-0-2-2:hover .switchOverlay-0-2-5 {
  opacity: 0.08;
}
.switchInput-0-2-3 {
  opacity: 0;
  position: absolute;
}
.switchInput-0-2-3:focus + .switchBase-0-2-2 .switchOverlay-0-2-5 {
  opacity: 0.12;
}
.switchTrack-0-2-4 {
  left: 10px;
  width: 36px;
  height: 14px;
  position: absolute;
  transition: background-color 150ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  border-radius: 20px;
}
.switchOverlay-0-2-5 {
  width: 40px;
  height: 40px;
  position: absolute;
  transition: left 150ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  border-radius: 50%;
}
.switchThumb-0-2-6 {
  width: 20px;
  height: 20px;
  position: absolute;
  box-shadow: 0px 1px 3px #00000036;
  transition: left 150ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  border-radius: 50%;
}
.switchBase-d0-0-2-7 {
  cursor: default;
  pointer-events: none;
}
.switchBase-d0-0-2-7:hover .switchOverlay-0-2-5 {
  background-color: #111111;
}
.switchInput-d1-0-2-8 {}
.switchInput-d1-0-2-8:focus + .switchBase-0-2-2 .switchOverlay-0-2-5 {
  background-color: #111111;
}
.switchTrack-d2-0-2-9 {
  background-color: #8c8c8c;
  opacity: 1;
}
.switchOverlay-d3-0-2-10 {
  left: 0;
}
.switchThumb-d4-0-2-11 {
  left: 10px;
  background-color: #ffffff;
}",
  <div>
    <span
      class=" switchContainer-0-2-1 Switch-gaya"
    >
      <input
        class="switchInput-0-2-3 switchInput-d1-0-2-8"
        disabled=""
        id="toggleSwitch"
        type="checkbox"
        value="test"
      />
      <label
        class="switchBase-0-2-2 switchBase-d0-0-2-7"
        for="toggleSwitch"
      >
        <span
          class="switchTrack-0-2-4 switchTrack-d2-0-2-9"
        />
        <span
          class="switchOverlay-0-2-5 switchOverlay-d3-0-2-10"
        />
        <span
          class="switchThumb-0-2-6 switchThumb-d4-0-2-11"
        />
      </label>
    </span>
  </div>,
]
`;

exports[`Switch component should render correctly when disabled is true and checked is true 1`] = `
[
  ".switchContainer-0-2-1 {
  width: 56px;
  height: 40px;
  display: flex;
  position: relative;
}
.switchBase-0-2-2 {
  width: 100%;
  height: 100%;
  display: flex;
  align-items: center;
}
.switchBase-0-2-2:hover .switchOverlay-0-2-5 {
  opacity: 0.08;
}
.switchInput-0-2-3 {
  opacity: 0;
  position: absolute;
}
.switchInput-0-2-3:focus + .switchBase-0-2-2 .switchOverlay-0-2-5 {
  opacity: 0.12;
}
.switchTrack-0-2-4 {
  left: 10px;
  width: 36px;
  height: 14px;
  position: absolute;
  transition: background-color 150ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  border-radius: 20px;
}
.switchOverlay-0-2-5 {
  width: 40px;
  height: 40px;
  position: absolute;
  transition: left 150ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  border-radius: 50%;
}
.switchThumb-0-2-6 {
  width: 20px;
  height: 20px;
  position: absolute;
  box-shadow: 0px 1px 3px #00000036;
  transition: left 150ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  border-radius: 50%;
}
.switchBase-d0-0-2-7 {
  cursor: default;
  pointer-events: none;
}
.switchBase-d0-0-2-7:hover .switchOverlay-0-2-5 {
  background-color: #f4ab34;
}
.switchInput-d1-0-2-8 {}
.switchInput-d1-0-2-8:focus + .switchBase-0-2-2 .switchOverlay-0-2-5 {
  background-color: #f4ab34;
}
.switchTrack-d2-0-2-9 {
  background-color: #8c8c8c;
  opacity: 1;
}
.switchOverlay-d3-0-2-10 {
  left: 16px;
}
.switchThumb-d4-0-2-11 {
  left: 26px;
  background-color: #ffffff;
}",
  <div>
    <span
      class=" switchContainer-0-2-1 Switch-gaya"
    >
      <input
        checked=""
        class="switchInput-0-2-3 switchInput-d1-0-2-8"
        disabled=""
        id="toggleSwitch"
        type="checkbox"
        value="test"
      />
      <label
        class="switchBase-0-2-2 switchBase-d0-0-2-7"
        for="toggleSwitch"
      >
        <span
          class="switchTrack-0-2-4 switchTrack-d2-0-2-9"
        />
        <span
          class="switchOverlay-0-2-5 switchOverlay-d3-0-2-10"
        />
        <span
          class="switchThumb-0-2-6 switchThumb-d4-0-2-11"
        />
      </label>
    </span>
  </div>,
]
`;
